dbms/src/Storages/Page/V3/PageStorageImpl.cpp

```diff
@@ -269,13 +279,139 @@ void PageStorageImpl::traverseImpl(const std::function<void(const DB::Page & pag
+                       full_gc_get_entries_ms,
+                       full_gc_blobstore_copy_ms,
+                       full_gc_apply_ms,
+                       get_external_msg());
+}
+
 bool PageStorageImpl::gcImpl(bool /*not_skip*/, const WriteLimiterPtr & write_limiter, const ReadLimiterPtr & read_limiter)
 {
     // If another thread is running gc, just return;
     bool v = false;
     if (!gc_is_running.compare_exchange_strong(v, true))
         return false;
 
+    const GCTimeStatistics consume = doGC(write_limiter, read_limiter);
+    assert(consume.stage != GCStageType::Unknown); // `doGC` must set the stage
+    LOG_DEBUG(log, consume.toLogging());
+
+    return consume.executeNextImmediately();
+}
+
+// Remove external pages for all tables
+// TODO: `clean_external_page` for all tables may slow down the whole gc process when there are lots of table.
+void PageStorageImpl::cleanExternalPage(Stopwatch & gc_watch, GCTimeStatistics & statistics)
+{
+    // Fine grained lock on `callbacks_mutex` and `callbacks_remove_mutex`.
+    // So that adding/removing a storage will not be blocked for the whole
+    // processing time of `cleanExternalPage`.
+    ExternalPageCallbacksContainer::iterator callbacks_iter;
+    {
+        std::scoped_lock lock{callbacks_mutex};
+        {
+            std::scoped_lock remove_lock{callbacks_remove_mutex};
+            // remove the callbacks by remove queue
+            for (auto ns_id : callbacks_remove_queue)
+                callbacks_container.erase(ns_id);
+            callbacks_remove_queue.clear();
+            callbacks_remove_queue.rehash(0);
+        }
+
+        // check and get the begin iter
+        statistics.num_external_callbacks = callbacks_container.size();
+        if (statistics.num_external_callbacks == 0)
+        {
+            statistics.clean_external_page_ms = gc_watch.elapsedMillisecondsFromLastTime();
+            return;
+        }
+        callbacks_iter = callbacks_container.begin();
+    }
+
+    SYNC_FOR("before_PageStorageImpl::cleanExternalPage_execute_callbacks");
+
+    Stopwatch external_watch;
+    // get all alive external ids from all namespaces
+    auto alive_external_ids = page_directory->getAliveExternalIds();
+    statistics.external_page_get_alive_ns = external_watch.elapsedFromLastTime();
+    while (true)
+    {
+        const auto & ns_id = callbacks_iter->first;
+        do
+        {
+            std::scoped_lock remove_lock{callbacks_remove_mutex};
+            // If the ns_id is pushed into `remove_queue`, then the `callbacks`
+            // are invalid and may cause failure, break do..while to continue
+            // on next namespace
+            if (callbacks_remove_queue.count(ns_id) > 0)
+                break;
+
+            // `remove_lock` ensure that the `callbacks` won't be invalid
+            const auto & callbacks = callbacks_iter->second;
+            auto pending_external_pages = callbacks.scanner();
```

Comment:
Do we need the protection of `remove_lock` after already having `callbacks`?

Reply:
The `this` ptr (DeltaMergeStore instance) inside callback will be released and cause crash if we run the 
lambda `scanner`/`remover`.

https://github.com/pingcap/tiflash/pull/5699/files#diff-30b0bd8d7bcbd8a75bdfa77186768765e94209ab4c3dedcac9da2ba54720695aR1272-R1303

Reply:
Can optimize this part? e.g. we can cache some information in the callback object, which is not strongly linked to DeltaMergeStore instance. So that the lock scope can be narrowed down to only put and get `callbacks_container` operations, and does not need to protect the `callbacks.scanner()` and `callbacks.remover()`

Reply:
And the lock mechanism will be much simpler, without two lock scopes.

Reply:
It is not that easy, in the callbacks we need the `StoragePathPool` instance, which the lifetime is the same as `DeltaMergeStore`, to remove useless DTFile 
https://github.com/pingcap/tiflash/blob/master/dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp#L316-L336

Reply:
If we want to decouple the callbacks with `DeltaMergeStore`, we may make the `StoragePathPool` to be a shared_ptr, and only save a weak_ptr to the `StoragePathPool` in the callbacks

Reply:
But even if we decouple the callbacks with `DeltaMergeStore`, we still need the lock `callbacks_remove_mutex`.
Because if thread A drop table that execute `callbacks.erase(ns_id)`, it invalid the iterator of ns_id. If thread B happen to hold the iterator of the same ns_id, thread B will crash.

We don't need that in `DeltaMergeStore.segments` and `PageDirctory.mvcc_table_directory` is that we won't execute `erase` in another thread

Reply:
Let's jump out of the details of the current code. The only reason we put a callback in PageStorage is that we want PS to clean up the useless DTFiles.
1. It does not matter that the callback is not put into `callbacks_container` in time, useless DTFiles are not required to be removed immediately.
2. Drop table will clean up the DTFiles of this table itself.

Back to the code, I don't understand why "thread A happen to hold the iterator of the same ns_id, thread A will crash".  As thread A should always hold a shared_ptr.

Reply:
The idea is trying to avoid blocking creating / deleting tables by GC, it will slow down the DDL performance.